### PR TITLE
[llvm-libgcc][CMake] Refactor llvm-libgcc

### PR DESCRIPTION
**Pull request by ur4t**
There are some issues in `llvm-libgcc` before this patch:

Commit c5a20b518203613497fa864867fc232648006068 ([llvm-libgcc] initial commit)
uses `$<TARGET_OBJECTS:unwind_static>` to get libunwind objects, which is empty.
The built library is actually a shared version of libclang_rt.builtins.

When configuring with `llvm/CMakeLists.txt`, target `llvm-libgcc` requires a
corresponding target in `llvm-libgcc/CMakeLists.txt`.

Per target installation is not handled by `llvm-libgcc`, which is not consistent
with `libunwind`.

This patch fixes those issues by:

Reusing target `unwind_shared` in `libunwind`, linking `compiler-rt.builtins`
objects into it, and applying version script.

Adding target `llvm-libgcc`, creating symlinks, and utilizing cmake's dependency
and component mechanism to ensure link targets will be built and installed along
with symlinks.

Mimicking `libunwind` to handle per target installation.

It is quite hard to set necessary options without further modifying the order of
runtime projects in `runtimes/CMakeLists.txt`. So though this patch reveals the
possibility of co-existence of `llvm-libgcc` and `compiler-rt`/`libunwind` in
`LLVM_ENABLE_RUNTIMES`, we still inhibit it to minimize influence on other
projects, considering that `llvm-libgcc` is only intended for toolchain vendors,
and not for casual use.

### CODE REVIEW

**Comment by compnerd** [comment]
CC: @cjdb

**Comment by MaskRay** [comment]
FWIW I did notice that the build had been broken (likely for a while) when I recently removed some `__{,de}register_frame_info*` and `__register_frame_table` in be91bd012123de835f64e10e677b24a7580b273c. Thank you for fixing the CMake.

```
-DLLVM_ENABLE_PROJECTS='clang;lld' -DLLVM_ENABLE_RUNTIMES='libcxx;libcxxabi;libunwind;llvm-libgcc;compiler-rt'
-DCOMPILER_RT_BUILTINS_HIDE_SYMBOLS=off  -DLLVM_LIBGCC_EXPLICIT_OPT_IN=on
```

This patch makes `libgcc` buildable but note that `llvm-libgcc` is not a target of the build directory
```
% ninja -C /tmp/out/custom libgcc
ninja: Entering directory `/tmp/out/custom'
ninja: error: unknown target 'libgcc
% ninja -C /tmp/out/custom llvm-libgcc
...
ninja: error: unknown target 'llvm-libgcc'
FAILED: runtimes/CMakeFiles/llvm-libgcc /tmp/out/custom/runtimes/CMakeFiles/llvm-libgcc
cd /tmp/out/custom/runtimes/runtimes-bins && /usr/bin/cmake --build /tmp/out/custom/runtimes/runtimes-bins/ --target llvm-libgcc --config Release
ninja: build stopped: subcommand failed.
% ninja -C /tmp/out/custom/runtimes/runtimes-bins/ libgcc  # good
% ninja -C /tmp/out/custom/runtimes/runtimes-bins/ llvm-libgcc
ninja: Entering directory `/tmp/out/custom/runtimes/runtimes-bins/'
ninja: error: unknown target 'llvm-libgcc'
```

Perhaps you can fix this as well.

**Comment by ur4t** [comment]
> This patch makes `libgcc` buildable but note that `llvm-libgcc` is not a target of the build directory

@MaskRay This issue is fixed in [c790fd0](https://github.com/llvm/llvm-project/pull/65455/commits/c790fd05b3cd4a5c5cbaff1d9855a8c4d7a1e2cb).

**Review by MaskRay — APPROVED** [review]

File: llvm-libgcc/CMakeLists.txt
Comment (cjdb) [inline llvm-libgcc/CMakeLists.txt:16]: Why is this being removed?

File: llvm-libgcc/CMakeLists.txt
Comment (cjdb) [inline llvm-libgcc/CMakeLists.txt:75]: This (and libunwind) need to be unconditional, since you shouldn't be building compiler-rt and libunwind independently when building llvm-libgcc.

File: llvm-libgcc/lib/blank.c
Comment (cjdb) [inline llvm-libgcc/lib/blank.c:1]: I'm happy you're able to delete this file.

**Review by cjdb — CHANGES_REQUESTED** [review]
Thanks for working on this!

IIRC it was @compnerd who convinced me that llvm-libgcc should **not** be compatible with an independent compiler-rt and libunwind, and that was because you only need one or the other. I still agree with that since llvm-libgcc is only intended for toolchain vendors, and not for casual use.

It would be good to get some exposition on why this PR changes this, and input from @compnerd regarding that design decision too.

**Comment by ur4t** [comment]
@cjdb @compnerd I am sorry for the misleading statement in this PR's initial comment.

> refactors `llvm-libgcc` to co-exist with `libunwind` and `compiler-rt`.

This statement shoule be refined: "refactors `llvm-libgcc` to co-exist with `libunwind` and `compiler-rt` *in `LLVM_ENABLE_RUNTIMES` list*".

> IIRC it was @compnerd who convinced me that llvm-libgcc should **not** be compatible with an independent compiler-rt and libunwind, and that was because you only need one or the other. I still agree with that since llvm-libgcc is only intended for toolchain vendors, and not for casual use.
> 
> It would be good to get some exposition on why this PR changes this, and input from @compnerd regarding that design decision too.

At first, I tried to build `llvm-libgcc` and found that unwind symbols in `libunwind.so` is not present in the built `libgcc_s.so`, and made [eb938ef](https://github.com/llvm/llvm-project/pull/65455/commits/eb938ef706ed7994fa0b362dfc4706e529a33563) to fix this issue.

Then I found that the original `llvm-libgcc/CMakeLists.txt`:
1. adds `compiler-rt` to get `libclang_rt.builtins-ARCH.a` object files.
2. adds `libunwind` to get `libunwind.a` object files.
3. replaces `libunwind.so` added in `libunwind` with a new `libunwind.so` built from object files in step 1 and step 2 and applies the version script.
4. installs replaced `libunwind.so` and versioned symlinks.
5. installs `libgcc` symlinks and versioned symlinks.

My first refactored `llvm-libgcc/CMakeLists.txt`:
1. adds `compiler-rt`.
2. adds `libunwind`.
3. modifies `libunwind.so` added in `libunwind` with `libclang_rt.builtins-ARCH.a` object files and applies the version script.
4. creates and installs `libgcc` symlinks and versioned symlinks.

Now `llvm-libgcc/CMakeLists.txt` keeps the original behaviour, but in a simplified approach, which means "`llvm-libgcc` is not compatible with an independent `compiler-rt` and `libunwind`" as before.

And I found the target clashes were no longer irreconcilable, via detecting `compiler-rt` and `libunwind` with `HAVE_COMPILER_RT` and `HAVE_LIBUNWIND` provided by `runtimes/CMakeLists.txt`, which allows `llvm-libgcc` to co-exist with `libunwind` and `compiler-rt` *in `LLVM_ENABLE_RUNTIMES` list*.

So there is no independent `compiler-rt` and `libunwind`, just removal of limitations while keeping the original behaviour.

File: llvm-libgcc/CMakeLists.txt
Comment (ur4t) [reply llvm-libgcc/CMakeLists.txt:75]: Yes, they are added unconditionally. `HAVE_COMPILER_RT` and `HAVE_LIBUNWIND` are set by `runtimes/CMakeLists.txt` if `LLVM_ENABLE_RUNTIMES` contains `compiler-rt` and `libunwind`. These checks are added to avoid target clashes when they are explicitly specified by users.

File: llvm-libgcc/CMakeLists.txt
Comment (ur4t) [reply llvm-libgcc/CMakeLists.txt:16]: `libunwind/CMakeLists.txt` does not contains a `project(libunwind LANGUAGES C CXX ASM)`. To keep consistent with `libunwind/CMakeLists.txt`, `project(llvm-libgcc LANGUAGES C CXX ASM)` is not added.

File: llvm-libgcc/lib/blank.c
Comment (ur4t) [reply llvm-libgcc/lib/blank.c:1]: In the refactored approach, we modify `libunwind.so` added in `libunwind` directly and no custom shared library is required, so the dummy source file can be deleted.

**Comment by ur4t** [comment]
@cjdb @compnerd With [b61aed7](https://github.com/llvm/llvm-project/pull/65455/commits/b61aed75b1c5891a4f84dd053fad1c68cc60902d), co-existence of `llvm-libgcc` and `libunwind`/`compiler-rt` in `LLVM_ENABLE_RUNTIMES` list is not allowed. Now I believe this PR is ready to be merged.

File: llvm-libgcc/CMakeLists.txt
Comment (ur4t) [reply llvm-libgcc/CMakeLists.txt:16]: Now `llvm-libgcc` use the same approach as `compiler-rt` in [60eccd0](https://github.com/llvm/llvm-project/pull/65455/commits/60eccd0b5b012e3e13bfdd4145dbd381a955887b).

Comment (cjdb) [reply llvm-libgcc/CMakeLists.txt:16]: I'm not following what you mean, sorry. Since this is the driver for compiler-rt and libunwind, it makes sense to me for this project to indicate this information.

Comment (ur4t) [reply llvm-libgcc/CMakeLists.txt:16]: `compiler-rt/CMakeLists.txt` detects whether `compiler-rt` is built as a subproject of `runtimes` or a standalone project. In standalone mode it sets options to detect `llvm` and declares the project name with `project(CompilerRT C CXX ASM)`.

Now `llvm-libgcc/CMakeLists.txt` does the same: in standalone mode, it tells added `compiler-rt` that we are not built as a subproject of `runtimes` and we have to detect `llvm` ourselves.

Here "standalone" means `cmake -B build -S compiler-rt` or `cmake -B build -S llvm-libgcc`, not `cmake -B build -S runtimes` nor `cmake -B build -S llvm`. Even in the whole `llvm-project` source tree, directly using `compiler-rt/CMakeLists.txt` or `llvm-libgcc/CMakeLists.txt` is "standalone".

If included by other `CMakeLists.txt`, `runtimes/CMakeLists.txt` for example, it does not matter whether there is a `project(CompilerRT C CXX ASM)` or `project(llvm-libgcc LANGUAGES C CXX ASM)` declaration actually performed. To keep consistent with `compiler-rt`, `llvm-libgcc/CMakeLists.txt` is adjusted correspondingly in [60eccd0](https://github.com/llvm/llvm-project/pull/65455/commits/60eccd0b5b012e3e13bfdd4145dbd381a955887b).

Correspondingly lines from [`compiler-rt/CMakeLists.txt`](https://github.com/llvm/llvm-project/blob/60eccd0b5b012e3e13bfdd4145dbd381a955887b/compiler-rt/CMakeLists.txt#L12) :
``` CMake
# Check if compiler-rt is built as a standalone project.
if (CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR OR COMPILER_RT_STANDALONE_BUILD)
  project(CompilerRT C CXX ASM)
  set(COMPILER_RT_STANDALONE_BUILD TRUE)
  set_property(GLOBAL PROPERTY USE_FOLDERS ON)
endif()
```

Correspondingly lines in [`llvm-libgcc/CMakeLists.txt`](https://github.com/llvm/llvm-project/blob/60eccd0b5b012e3e13bfdd4145dbd381a955887b/llvm-libgcc/CMakeLists.txt#L9):
``` CMake
# Check if llvm-libgcc is built as a standalone project
if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR OR LLVM_LIBGCC_STANDALONE_BUILD)
  project(llvm-libgcc LANGUAGES C CXX ASM)
  set(COMPILER_RT_STANDALONE_BUILD ON)
  set_property(GLOBAL PROPERTY USE_FOLDERS ON)
  set(LLVM_LIBGCC_COMPILER_RT_BINARY_DIR "compiler-rt")
  set(LLVM_LIBGCC_LIBUNWIND_BINARY_DIR "libunwind")
else()
  set(LLVM_LIBGCC_COMPILER_RT_BINARY_DIR "../compiler-rt")
  set(LLVM_LIBGCC_LIBUNWIND_BINARY_DIR "../libunwind")
endif()
```

And I am a bit confused how should we indicate "this is the driver for `compiler-rt` and `libunwind`". It is completely acceptable to perform `project(llvm-libgcc LANGUAGES C CXX ASM)` declaration unconditionally, but does it makes sense in this scenario?

File: llvm-libgcc/CMakeLists.txt
Comment (cjdb) [reply llvm-libgcc/CMakeLists.txt]: The design is that you _should not_ be adding `compiler-rt` or `libunwind` to `LLVM_ENABLE_RUNTIMES`. Why is this being changed?

File: llvm-libgcc/CMakeLists.txt
Comment (cjdb) [reply llvm-libgcc/CMakeLists.txt:16]: Understood, thanks for highlighting the change to [llvm-libgcc/CMakeLists.txt](https://github.com/llvm/llvm-project/blob/60eccd0b5b012e3e13bfdd4145dbd381a955887b/llvm-libgcc/CMakeLists.txt#L9), which I'd missed.

File: llvm-libgcc/CMakeLists.txt
Comment (ur4t) [reply llvm-libgcc/CMakeLists.txt]: It was changed, and now it is restored in [b61aed7](https://github.com/llvm/llvm-project/pull/65455/commits/b61aed75b1c5891a4f84dd053fad1c68cc60902d).

I tried to make them co-exist happily and found that it is quite hard to set necessary options without further modifying the order of runtime projects in `runtimes/CMakeLists.txt`.

This finding is documented in the commit message to explain why we chose this approach.

**Review by cjdb — APPROVED** [review]
Thanks for the patch!

**Comment by ldionne** [comment]
Are we waiting on anything to merge this? I think folks' expectations of how patches are merged might be unclear right now due to the Github PR transition. With Phabricator the patch author would push the change and the "PR" would be closed automatically. Now with Github PRs someone has to press "squash and merge" but it might not be clear whose responsibility it is to do this.

@cjdb @MaskRay We're ready to land this, right? If so, I think either of you can "Squash and merge" (or @ur4t if they are able to).

**Comment by cjdb** [comment]
Right, it's ready to land. Thanks for pointing out that the workflow is different now.

**Comment by cjdb** [comment]
@ur4t before I squash and merge, could you please provide me with a commit message that explains why the changes were necessary (as a comment)? I've looked at all the commit messages and text, but I don't see anything that captures the full message. If you've already written something, a link to that will be fine :)

**Comment by MaskRay** [comment]
I agree with a better description is needed.

> Are we waiting on anything to merge this? I think folks' expectations of how patches are merged might be unclear right now due to the Github PR transition. With Phabricator the patch author would push the change and the "PR" would be closed automatically. Now with Github PRs someone has to press "squash and merge" but it might not be clear whose responsibility it is to do this.

I feel the author is the best one to perform the action, as they can adjust the description to be clearer, rebase as needed, or do other minor adjustments that do not need re-review.

**Comment by ldionne** [comment]
> 
> I feel the author is the best one to perform the action, as they can adjust the description to be clearer, rebase as needed, or do other minor adjustments that do not need re-review.

FWIW that may not always be possible for folks who don't have direct commit access to the repo. But feel free to handle that however makes the most sense to you. For libc++/libc++abi, I often merge on behalf of folks to make sure that the PR isn't forgotten, etc.

**Comment by ur4t** [comment]
@cjdb Sorry for the circuitous commit history. I am glad to provide a summary.


[llvm-libgcc][CMake] Refactor llvm-libgcc

There are some issues in `llvm-libgcc` before this patch:

Commit c5a20b518203613497fa864867fc232648006068 ([llvm-libgcc] initial commit)
uses `$<TARGET_OBJECTS:unwind_static>` to get libunwind objects, which is empty.
The built library is actually a shared version of libclang_rt.builtins.

When configuring with `llvm/CMakeLists.txt`, target `llvm-libgcc` requires a
corresponding target in `llvm-libgcc/CMakeLists.txt`.

Per target installation is not handled by `llvm-libgcc`, which is not consistent
with `libunwind`.


This patch fixes those issues by:

Reusing target `unwind_shared` in `libunwind`, linking `compiler-rt.builtins`
objects into it, and applying version script.

Adding target `llvm-libgcc`, creating symlinks, and utilizing cmake's dependency
and component mechanism to ensure link targets will be built and installed along
with symlinks.

Mimicking `libunwind` to handle per target installation.


It is quite hard to set necessary options without further modifying the order of
runtime projects in `runtimes/CMakeLists.txt`. So though this patch reveals the
possibility of co-existence of `llvm-libgcc` and `compiler-rt`/`libunwind` in
`LLVM_ENABLE_RUNTIMES`, we still inhibit it to minimize influence on other
projects, considering that `llvm-libgcc` is only intended for toolchain vendors,
and not for casual use.

**Comment by ur4t** [comment]
@MaskRay It seems that Github enforce a column limit of 70 characters in "Squash and Merge" commit messages ([refined-github#1475](https://github.com/refined-github/refined-github/issues/1475)). Commit messages targeting a 80 character limitation will be truncated again when texts are copied and pasted directly with original line breakages (7ac1f60f01a73f9678bba2385e933864ecce8622 and af935cf0ee8a26dcf0860ed0c72d74b4a1cb1314). Should we mention this behaviour in documents for the Github PR transition?

**Comment by MaskRay** [comment]
> @MaskRay It seems that Github enforce a column limit of 70 characters in "Squash and Merge" commit messages ([refined-github#1475](https://github.com/refined-github/refined-github/issues/1475)). Commit messages targeting a 80 character limitation will be truncated again when texts are copied and pasted directly with original line breakages ([7ac1f60](https://github.com/llvm/llvm-project/commit/7ac1f60f01a73f9678bba2385e933864ecce8622) and [af935cf](https://github.com/llvm/llvm-project/commit/af935cf0ee8a26dcf0860ed0c72d74b4a1cb1314)). Should we mention this behaviour in documents for the Github PR transition?

Thanks for mentioning this behavior. Perhaps comment on https://discourse.llvm.org/t/hows-it-going-with-pull-requests/73467 or a related github pr transition post?

Due to email spam from github notifications, I almost missed your comment...

**Pull request merged**
